lib/unparser/preprocessor.rb

```diff
@@ -151,7 +151,57 @@ def chunked_children
 
     end # CollapseStrChildren
 
-    # Preprocessor eliminating unneded dstr nodes
+    # Preprocessor flattening unneeded dstr nesting. This happens when a dstr is generated from
+    # implicit string concatenation, and one of the strings being concatenated is itself a dstr.
+    #
+    # Example: "foo" "bar #{baz}"
+    #
+    # Without this preprocessor, this turns into: "foo#{"bar#{baz}"}"
+    # With this preprocessor, this turns into: "foobar #{baz}"
+    class FlattenImplicitDSTR < self
+
+      # No need to register :dsym, because Ruby doesn't do implicit concatenation of symbols
+      register :dstr
+
+      FLATTEN_CHILDREN = IceNine.deep_freeze([:dstr, :str])
+
+      # Return preprocessor result
+      #
+      # @return [Parser::AST::Node]
+      #
+      # @api private
+      #
+      def result
+        return node unless implicit_dstr?
+
+        flat_children = children.flat_map do |child|
+          if child.type.equal?(:str)
```

Comment:
Instead of doing nothing for `:str`, and doing something explicit for `:dstr`, would it be safer to match `:dstr` explicitly and do the work, otherwise defaulting to a no-op? Maybe it's impossible for the node in that location to be something other than `:str` or `:dstr`, but it would be more future proof written as:

```ruby
if child.type.equal?(:dstr)
  child.children
else
  child
end
```

Since it's a simple conditional, it could probably be written as a ternary:

```ruby
child.type.equal?(:dstr) ? child.children : child
```